ui/oopsyraidsy/data/05-shb/raid/e7n.js

```diff
@@ -6,9 +6,81 @@
     ko: /^희망의 낙원 에덴: 공명편 \(3\)$/,
   },
   damageWarn: {
+    '4C55': 'Stygian Sword', // Circle ground AoEs after False Twilight
+    '4C4C': 'Strength In Numbers 1', // Large donut ground AoEs, intermission
+    '4C4D': 'Strength In Numbers 2', // Large circle ground AoEs, intermission
   },
   damageFail: {
   },
   triggers: [
+    {
+      id: 'E7N Stygian Stake', // Laser tank buster, outside intermission phase
+      damageRegex: '4C33',
+      condition: function(e) {
+        return e.type != '15';
+      },
+      mistake: function(e) {
+        return { type: 'warn', blame: e.targetName, text: e.abilityName };
+      },
+    },
+    {
+      id: 'E5N Silver Shot', // Spread markers, intermission
+      damageRegex: '4E7D',
+      condition: function(e) {
+        return e.type != '15';
+      },
+      mistake: function(e) {
+        return { type: 'warn', blame: e.targetName, text: e.abilityName };
+      },
+    },
+    {
+      // If we don't set this up ahead of time, the Dark/Light Course triggers will fail
+      // on the first two uses of Light's Course, since the Astral/Umbral properties wouldn't exist.
+      id: 'E7N Debuff Setup',
+      regex: Regexes.startsUsing({ source: 'The Idol Of Darkness', id: '4C2B' }),
+      condition: function(e, data) {
+        return !(data.hasAstral && data.hasUmbral);
+      },
+      run: function(e, data) {
+        data.hasAstral = data.hasAstral || {};
+        data.hasUmbral = data.hasUmbral || {};
+      },
+    },
+    {
+      id: 'E7N Astral Tracking',
+      gainsEffectRegex: gLang.kEffect.AstralEffect,
+      losesEffectRegex: gLang.kEffect.AstralEffect,
+      run: function(e, data) {
+        data.hasAstral[e.targetName] = e.gains;
+      },
+    },
+    {
+      id: 'E7N Umbral Tracking',
+      gainsEffectRegex: gLang.kEffect.UmbralEffect,
+      losesEffectRegex: gLang.kEffect.UmbralEffect,
+      run: function(e, data) {
+        data.hasUmbral[e.targetName] = e.gains;
+      },
+    },
+    {
+      id: 'E7N Light\'s Course',
+      damageRegex: ['4C3E', '4C40', '4C22', '4C3C', '4E63'],
+      condition: function(e, data) {
+        return !data.hasUmbral[e.targetName];
+      },
+      mistake: function(e) {
+        return { type: 'fail', blame: e.targetName, text: e.abilityName + ' wrong/no buff' };
```

Comment:
I'm confused at the `/ no buff` part here?

Also, why are there more ids in light's course vs dark's course? Is one of these the initial uncolored ones, and those should be up in the damageFail category?

Reply:
The same IDs are used for Light's Course whether or not the players are inflicted with Astral/Umbral, at any phase of the battle. It is *technically* not a failure to be hit during a non-debuff phase, in the sense that the player will not auto-die. What if we did something like this?

```
    {
      id: 'E7N Light\'s Course',
      damageRegex: ['4C3E', '4C40', '4C22', '4C3C', '4E63'],
      condition: function(e, data) {
        return !data.hasUmbral[e.targetName];
      },
      mistake: function(e) {
        if (data.hasAstral[e.targetName])
          return { type: 'fail', blame: e.targetName, text: e.abilityName + ' wrong buff' };
        return { type: 'warn', blame: e.targetName, text: e.abilityName + ' no buff' };
      },
    },
```

As for the disparate numbers of IDs, I'm not sure, but I have a theory, It *looks* like each Course ability hits two "lanes", while the actual areas hit on the battlefield are dependent on which Unforgiven Idolatry uses which ability. (I base this supposition on the fact that the first two sets of birds for the encounter are always fixed, with the outer lanes being hit before inner with Light's course. Both of these hits, inner and outer, use the `4C3C` cast.) The "extra" ID comes from the unbalanced set of Light's Course casts at the start.

(We could *probably* dig into that more and use it to figure out some useful positioning triggers for Savage? I don't know.)

A quick search at XIVAPI shows 13 different versions of Light's Course compared with 11 for Dark. I do know that across my logs I've only seen the IDs I've put in the oopsy file.